### PR TITLE
Support Multiple Hosts for a Route

### PR DESCRIPTION
**Pull request by Kahbazi**
Fixes https://github.com/microsoft/reverse-proxy/issues/206

### CODE REVIEW

File: docs/docfx/articles/getting_started.md [hunk 1]
```diff
@@ -106,15 +106,15 @@ You can find out more about the available configuration options by looking at [P
         "ClusterId": "cluster1",
         "Match": {
           "Methods": [ "GET", "POST" ],
-          "Host": "localhost",
+          "Host": [ "localhost" ],
```
Comment (Tratcher) [inline docs/docfx/articles/getting_started.md:109]:
```suggestion
          "Hosts": [ "localhost" ],
```
In all config files

File: src/ReverseProxy/Service/DynamicEndpoints/RuntimeRouteBuilder.cs [hunk 2]
```diff
@@ -68,9 +69,9 @@ public RouteConfig Build(ParsedRoute source, ClusterInfo cluster, RouteInfo runt
             endpointBuilder.DisplayName = source.RouteId;
             endpointBuilder.Metadata.Add(newRouteConfig);
 
-            if (!string.IsNullOrEmpty(source.Host))
+            if (source.Hosts != null && !source.Hosts.All(host => string.IsNullOrEmpty(host)))
```
Comment (Tratcher) [inline src/ReverseProxy/Service/DynamicEndpoints/RuntimeRouteBuilder.cs:72]: Empty ones should have been rejected before this. Here you should be able to check Hosts.Count  != 0 now.

File: src/ReverseProxy/Service/Config/RuleParsing/RouteValidator.cs [hunk 3]
```diff
@@ -68,13 +69,13 @@ public async Task<bool> ValidateRouteAsync(ParsedRoute route, IConfigErrorReport
                 success = false;
             }
 
-            if (string.IsNullOrEmpty(route.Host) && string.IsNullOrEmpty(route.Path))
+            if ((route.Hosts == null || route.Hosts.All(host => string.IsNullOrEmpty(host))) && string.IsNullOrEmpty(route.Path))
```
Comment (Tratcher) [inline src/ReverseProxy/Service/Config/RuleParsing/RouteValidator.cs:72]: Hosts.Count == 0 || Hosts.Any(host => string.IsNullOrEmpty(host))...

File: src/ReverseProxy/Service/Config/RuleParsing/RouteValidator.cs [hunk 4]
```diff
@@ -83,18 +84,21 @@ public async Task<bool> ValidateRouteAsync(ParsedRoute route, IConfigErrorReport
             return success;
         }
 
-        private static bool ValidateHost(string host, string routeId, IConfigErrorReporter errorReporter)
+        private static bool ValidateHost(IReadOnlyList<string> hosts, string routeId, IConfigErrorReporter errorReporter)
         {
             // Host is optional when Path is specified
-            if (string.IsNullOrEmpty(host))
+            if (hosts == null || hosts.All(host => string.IsNullOrEmpty(host)))
```
Comment (Tratcher) [inline src/ReverseProxy/Service/Config/RuleParsing/RouteValidator.cs:90]: Any empty array would constitute no hosts (equivalent to the prior IsNullOrEmpty check), but I think any null or empty values in the array are an error.
hosts.Count == 0

File: src/ReverseProxy/Service/Config/RuleParsing/RouteValidator.cs [hunk 5]
```diff
@@ -83,18 +84,21 @@ public async Task<bool> ValidateRouteAsync(ParsedRoute route, IConfigErrorReport
...
             {
                 return true;
             }
 
-            if (!_hostNameRegex.IsMatch(host))
+            for (var i = 0; i < hosts.Count; i++)
             {
-                errorReporter.ReportError(ConfigErrors.ParsedRouteRuleInvalidMatcher, routeId, $"Invalid host name '{host}'");
-                return false;
+                if (!_hostNameRegex.IsMatch(hosts[i]))
```
Comment (Tratcher) [inline src/ReverseProxy/Service/Config/RuleParsing/RouteValidator.cs:97]: string.IsNullOrEmpty(hosts[i]) || ...

**Comment by Kahbazi** [comment]
@Tratcher Any Idea why the build is failing? It builds successfully on my machine. I can't figure out the reason. The error is `'ParsedRoute' does not contain a definition for 'Host'` .

**Comment by Tratcher** [comment]
It's merging with the CORS changes before building. You need to rebase.

**Comment by Tratcher** [comment]
Thanks

**Pull request merged**